api/server/router/image/image_routes.go

```diff
@@ -415,6 +416,11 @@ func (ir *imageRouter) postImagesTag(ctx context.Context, w http.ResponseWriter,
 		return errdefs.InvalidParameter(err)
 	}
 
+	refName := reference.FamiliarName(ref)
+	if refName == string(digest.Canonical) {
+		return errdefs.InvalidParameter(errors.New("refusing to create an ambiguous tag using digest algorithm as name"))
+	}
```

Comment:
This change looks good (and probably similar to what we did/do in the non-containerd version?). 


We should look at follow-ups though, as ideally we'd take _other_ algorithms into account as well, as they _may_ occur as well.

Let me write it down here, but I'll open a separate ticket from it;

The spec defines a couple of algorithms that are _registered_ in the spec (sha256, sha512);
https://github.com/opencontainers/image-spec/blob/v1.1.0-rc5/descriptor.md#registered-algorithms, so those are good candidates to _always_ ignore (as the spec defines that the can be used).

It also describes some other (more esoteric) ones as "optional" https://github.com/opencontainers/image-spec/blob/v1.1.0-rc5/descriptor.md#digests

And potentially other algorithms (`sha384`, and `blake3` variants) as well;

- https://github.com/opencontainers/go-digest/blob/ea51bea511f75cfa3ef6098cc253c5c3609b037a/algorithm.go#L33-L35
- https://github.com/opencontainers/go-digest/pull/66
- https://github.com/opencontainers/image-spec/issues/819
- :warning: all of which could be "dynamically" registered: https://github.com/opencontainers/go-digest/pull/62
- :question: :warning: and _possibly_ even other ones that are implicitly registered (looking at you, `md5`)

So, I think we may need some follow-ups;

- Perhaps `go-digest` could have a utility function that returns the list of registered algorithms (assuming that's the canonical list of algorithms that we accept: https://github.com/opencontainers/go-digest/blob/316f76db9b0f1b3620c3380cc34f75d262dea6db/algorithm.go#L78-L96)
- A similar utility could/should be added to https://github.com/distribution/reference. Baring an OCI specification for image-references, I think we should consider that to be the canonical implementation for image references
- So it should _at least_ have a utility to check for ambiguous references
- And _possibly_ invalidate such references automatically
- for any "image reference spec", it would be hard to define a "dynamic" list though. and unfortunately, [the `go-digest` regex is overly broad](https://github.com/opencontainers/go-digest/blob/316f76db9b0f1b3620c3380cc34f75d262dea6db/algorithm.go#L42), so can't be used to detect whether we're dealing with a (possible) algorithm, or an image name (it looks like some context on that RegEx was removed in https://github.com/opencontainers/go-digest/commit/b9e02e015be61903bbee58e3fd349114fa28e0b4 :thinking:)
- So _maybe_ such a spec must define a minimum set of algorithms to consider "reserved" (MUST be rejected), and possibly a recommendation on what patterns to _consider_ an algorithm (which MAY be rejected).

Reply:
- opened https://github.com/moby/moby/issues/46500